aws/resource_aws_xray_sampling_rule.go

```diff
@@ -0,0 +1,212 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/xray"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsXraySamplingRule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsXraySamplingRuleCreate,
+		Read:   resourceAwsXraySamplingRuleRead,
+		Update: resourceAwsXraySamplingRuleUpdate,
+		Delete: resourceAwsXraySamplingRuleDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"rule_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 128),
+			},
+			"resource_arn": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"priority": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"fixed_rate": {
+				Type:     schema.TypeFloat,
+				Required: true,
+			},
+			"reservoir_size": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"service_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"service_type": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"host": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"http_method": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"url_path": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"version": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"attributes": {
```

Comment:
Two notes:

* While Terraform does not explicitly error during resource schema validation without it currently, we prefer to add the explicit `Elem` declaration to delineate the underlying map value type, as it likely will when a future version of the Terraform Provider SDK supports multiple map value types. I don't see an explicit `Elem` section in the [Extending Terraform documentation section on schemas](https://www.terraform.io/docs/extend/schemas/index.html) so I will make a note to ensure this gets added.
* We can add validation for this attribute's keys and values to match the [API reference](https://docs.aws.amazon.com/xray/latest/api/API_SamplingRule.html) via:

```go
Elem: &schema.Schema{
	Type: schema.TypeString,
	ValidateFunc: validation.StringLenBetween(1, 32),
},
ValidateFunc: validation.StringLenBetween(1, 32),
```

Reply:
Added the `Elem` part -  However, `ValidateFunc: validation.StringLenBetween(1, 32),` on the TypeMap level causes this:
Input:
```hcl
...
     attributes = {
         Hello = "World"
     }
...
```
Output:
`testing.go:568: Step 0 error: config is invalid: expected type of attributes to be string`